File: services/api/routers/code.py
from __future__ import annotations

from fastapi import APIRouter

from services.api.schemas import (
    LspCompleteOut,
    LspHoverOut,
    LspPositionIn,
    SyntaxCheckIn,
    SyntaxCheckOut,
)
from services.api.services.lsp import complete_python, hover_python
from services.api.services.syntax import check_python

router = APIRouter(prefix="/code", tags=["code"])


@router.post("/syntax", response_model=SyntaxCheckOut)
def python_syntax(payload: SyntaxCheckIn) -> SyntaxCheckOut:
    return check_python(payload.code)


@router.post("/complete", response_model=LspCompleteOut)
def python_complete(payload: LspPositionIn) -> LspCompleteOut:
    return complete_python(payload.code, payload.line, payload.column)


@router.post("/hover", response_model=LspHoverOut)
def python_hover(payload: LspPositionIn) -> LspHoverOut:
    return hover_python(payload.code, payload.line, payload.column)

File: apps/web/src/lib/api/code.ts
import { request } from "./http";
import type { LspCompletion } from "./types";

export const codeApi = {
  checkSyntax: (code: string) =>
    request<{
      ok: boolean;
      message: string | null;
      line: number | null;
      column: number | null;
    }>("/api/v1/code/syntax", {
      method: "POST",
      body: JSON.stringify({ code }),
    }),
  completePython: (code: string, line: number, column: number) =>
    request<{
      items: LspCompletion[];
      syntax: { ok: boolean };
      error?: string | null;
    }>("/api/v1/code/complete", {
      method: "POST",
      body: JSON.stringify({ code, line, column }),
    }),
  hoverPython: (code: string, line: number, column: number) =>
    request<{ contents: string | null; error?: string | null }>(
      "/api/v1/code/hover",
      {
        method: "POST",
        body: JSON.stringify({ code, line, column }),
      },
    ),
};
